Add regex validation for the entity id filter

flip-dots opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

This adds regex validation for the filter_entity_id parameter in the history part of the REST API to make sure that the provided entity IDs are valid. Previously the API would just return an empty array instead of an error message, which is (probably) not expected behaviour.

This makes it easier to spot if you made a mistake in your request formatting, as you now get a specific error instead of an empty array which could be confusing.

Also for the WebSocket API

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

flip-dots committed:
Add regex validation for the entity id filter

bdraco commented on homeassistant/components/history/__init__.py:
```diff
@@ -120,6 +120,11 @@ async def get(
         if entity_ids_str:
             entity_ids = entity_ids_str.lower().split(",")
+            for entity_id in entity_ids:
+                if not valid_entity_id(entity_id):
```
Please check if it's in the state machine first before calling this check as it can get expensive if the entity id list is large

Also this needs a test.

Thanks

bdraco replied:
```suggestion
                if not hass.states.get(entity_id) or not valid_entity_id(entity_id):
```

flip-dots committed:
Add test for history API entity_id filter

flip-dots committed:
Add regex validation for entity_ids in WS API

flip-dots committed:
Add tests for entity_ids in history WS API

flip-dots committed:
Add check to see if the entity_id(s) exist

Co-authored-by: J. Nick Koston <nick@koston.org>

flip-dots committed:
Add check to see if the entity_id(s) exist for WS API

flip-dots committed:
Remove feature causing failing test

This was causing some of the tests to fail, I guess because in some circumstances it might be desirable to subscribe to the state stream of a non-existent entity, so its probably best to just check to make sure that the entitiy_id is valid and not care if it doesnt exist.

flip-dots commented on homeassistant/components/history/__init__.py:
```diff
@@ -120,6 +120,11 @@ async def get(
         entity_ids = None
         if entity_ids_str:
             entity_ids = entity_ids_str.lower().split(",")
+            for entity_id in entity_ids:
+                if not hass.states.get(entity_id) or not valid_entity_id(entity_id):
+                    return self.json_message(
+                        "Invalid filter_entity_id", HTTPStatus.BAD_REQUEST
+                    )
```
Now that I think about it, the regex check in the REST API is redundant when we also check the state machine, since I think we can assume that if an entitiy_id is in the state machine it must be valid, so doing a regex check only makes sense for the Web Socket API where you can subscribe to entity_ids which don't exist yet.
